Add Ttl value object for cache time-to-live

Pekhov14 opened the pull request:
| Q             | A
| ------------- | ---
| Is bugfix?    | ❌
| New feature?  | ✔️
| Breaks BC?    | ❌
| Fixed issues  | ❌

## This PR introduces a new immutable value object `Ttl` that represents cache time-to-live (TTL) in seconds.
Its main goals are:

- Improve readability — remove magic numbers like `60 * 60`.
- Ensure type safety — TTLs are now strongly typed.
- Provide convenience factories for common units: seconds(), minutes(), hours(), days().
- Normalize TTL values consistently via a single entry point: Ttl::from().

Currently, cache TTLs are often passed as raw integers (seconds), e.g.:
```php
$cache = new \Yiisoft\Cache\Cache($arrayCache, 60 * 60); // 1 hour
$cache->set('key', 'value', 3600); // 1 hour
```

With `Ttl`, all TTL values become self-explanatory and type-safe.
Immutable — once created, the value cannot be changed.

### Convenient factories
```php
Ttl::seconds(30);   // 30 seconds
Ttl::minutes(15);   // 15 minutes
Ttl::hours(2);      // 2 hours
Ttl::days(1);       // 1 day

// Complex durations
$ttl = Ttl::create(sec: 30, min: 10, hour: 1); // 1 hour 10 minutes 30 seconds
$cache->set('key', 'value', $ttl->toSeconds());

// From DateInterval
$ttl = Ttl::fromInterval(new DateInterval('PT45M'));
$cache->set('key', 'value', $ttl);
```

### Normalization via Ttl::from()
Ttl::from() serves as a unified TTL normalizer. It accepts multiple formats:
```php
Ttl::from(3600);                      // integer seconds
Ttl::from('3600');                    // string seconds
Ttl::from(new DateInterval('PT45M')); // DateInterval
Ttl::from($ttlInstance);              // Ttl instance
Ttl::from(null);                      // infinity / no expiration

Ttl::forever()                        // infinity / no expiration
```
This ensures that all cache TTL values, no matter their original type, can be consistently converted to a `Ttl` object internally.

### Integration with cache API
```php
$cache->set('key', 'value', Ttl::minutes(15)->toSeconds());
$cache->set('key', 'value', Ttl::create(hour:5, min:30)->toSeconds());

// Or using a normalized TTL object
$ttl = Ttl::from(new DateInterval('PT45M'));
$cache->set('key', 'value', $ttl?->toSeconds());
```
### Benefits
- Self-documenting code — TTL durations are explicit and readable.
- Consistent handling of TTLs across different cache backends.
- Easier maintenance — no more magic numbers sprinkled across the codebase.
- BC-safe — you can still pass integers or `DateIntervals`; `Ttl` provides optional normalization.

Pekhov14 commented:
I did not change your tests/TestCase.php; your CI did. 🙂

vjik commented on src/Ttl.php:
```diff
@@ -0,0 +1,60 @@
+        return new self($day * self::SECONDS_IN_DAY);
+    }
+
+    public function toInt(): int
```
```suggestion
    public function toSeconds(): int
```

vjik commented on tests/TtlTest.php:
```diff
@@ -0,0 +1,36 @@
+    #[\PHPUnit\Framework\Attributes\DataProvider('ttlProvider')]
+    public function testFactories(Ttl $ttl, int $expectedSeconds): void
+    {
+        $this->assertInstanceOf(Ttl::class, Ttl::seconds(10));
+        $this->assertInstanceOf(Ttl::class, Ttl::minutes(5));
+        $this->assertInstanceOf(Ttl::class, Ttl::hours(2));
+        $this->assertInstanceOf(Ttl::class, Ttl::days(1));
+        $this->assertInstanceOf(Ttl::class, Ttl::create(sec:1, min:1, hour:1, day:1));
```
Extract to separate test method

vjik commented on src/Ttl.php:
```diff
@@ -0,0 +1,60 @@
+        return $this->value;
+    }
+
+    public function __invoke(): int
```
Seems, `toSeconds()` method is enough and `__invoke()` can be removed.

Pekhov14 replied:
@vjik Okay, I agree. I just thought it would be useful for shorter expressions.

Pekhov14 replied:
Okay, I agree. I just thought it would be useful for shorter expressions.

vjik commented on src/Ttl.php:
```diff
@@ -0,0 +1,60 @@
+
+namespace Yiisoft\Cache;
+
+final class Ttl
```
Add phpdoc for class and his methods.

vjik replied:
Add line to changelog

vjik commented on src/Ttl.php:
```diff
@@ -0,0 +1,60 @@
+
+namespace Yiisoft\Cache;
+
+final class Ttl
```
Should we use this object for TTL parameters in other classes in this package?

vjik replied:
We can add new class usage without break BC, for example:

```php
public function __construct(\Psr\SimpleCache\CacheInterface $handler, DateInterval|int|null $defaultTtl = null)
// change to
public function __construct(\Psr\SimpleCache\CacheInterface $handler, Ttl|DateInterval|int|null $defaultTtl = null)
```

Also we can into classes cast all ttl values to `Ttl` object and use it inside already now.
 
By the way, method `fromInterval(DateInterval $interval)` is should be added.

samdark commented:
Overall, it looks like a good syntactic sugar :+1:

samdark commented on README.md:
```diff
@@ -66,6 +66,33 @@ Set a default TTL:
+use Yiisoft\Cache\Ttl;
+
+// Simple usage
+$cache->set('key', 'value', Ttl::seconds(30)->toSeconds()); // 30 seconds
+$cache->set('key', 'value', Ttl::minutes(15)->toSeconds()); // 15 minutes
+$cache->set('key', 'value', Ttl::hours(2)->toSeconds());    // 2 hours
+$cache->set('key', 'value', Ttl::days(1)->toSeconds());     // 1 day
+
+// Complex durations
+$ttl = Ttl::create(sec: 30, min: 10, hour: 1); // 1 hour 10 minutes 30 seconds
+$cache->set('key', 'value', $ttl->toSeconds());
+
+// From DateInterval
+$ttl = Ttl::from(new DateInterval('PT45M')); // 45 minutes
+# OR
+$ttl = Ttl::fromInterval(new DateInterval('PT45M'));
+$cache->set('key', 'value', $ttl);
+
+// Infinity (no expiration)
+$ttl = Ttl::from(null);
+$cache->set('key', 'value', $ttl?->toSeconds());
```
Do we need to explicitly call `toSeconds()` here?

Pekhov14 replied:
@samdark 
For `ArrayCache` and `NullCache` there is no need to call `toSeconds()`, since they already extend the signature and accept `Ttl` directly.
For other adapters, I can add it after this PR is approved.

Do you think it’s worth mentioning in the README that for these adapters (`ArrayCache` and `NullCache`) `toSeconds()` is not required?

samdark replied:
What's in the README will be used exactly.

Pekhov14 commented:
Hi, I’m submitting the current code as-is. Could you advise on how I could add `Ttl` support for other adapters?

Right now, in their composer.json, the adapters don’t depend on the main yiisoft/cache package, so specifying `Ttl` as a type isn’t possible.

Would it make sense to extract `Ttl` into a separate package (e.g., yiisoft/ttl) and then integrate it into yiisoft/cache and all other adapters? Or is the current approach sufficient?

samdark commented:
Adapters are PSR drivers. They are not supposed to support TTL directly. Cache wrapper should accept TTL though.

Pekhov14 commented:
> Adapters are PSR drivers. They are not supposed to support TTL directly. Cache wrapper should accept TTL though.

Refactored `ArrayCache` and `NullCache` so public methods no longer accept `Ttl` directly. Internally, Ttl::from() is still used to normalize TTL values, keeping the implementation consistent while remaining fully PSR-compliant. 
This keeps TTL handling convenient without affecting adapter contracts.

Pekhov14 commented on README.md:
```diff
@@ -66,6 +66,36 @@ Set a default TTL:
+$cache->set('key', 'value', Ttl::seconds(30)->toSeconds()); // 30 seconds
+$cache->set('key', 'value', Ttl::minutes(15)->toSeconds()); // 15 minutes
+$cache->set('key', 'value', Ttl::hours(2)->toSeconds());    // 2 hours
+$cache->set('key', 'value', Ttl::days(1)->toSeconds());     // 1 day
```
Thanks for the comment, @samdark! 🙏

You're right — I should clarify which cache instance is being used. In this example, $cache is a `PSR-16` cache adapter (`ArrayCache`), not the Yii Cache wrapper, which doesn’t have a `set()` method.

I've updated the README to make it clear which type of cache is being used in each example — hopefully, it’s more clear now.

Pekhov14 commented on README.md:
```diff
@@ -66,6 +66,36 @@ Set a default TTL:
+// Infinity (no expiration)
+$cache->set('key', 'value', Ttl::from(null));
+# OR
+$cache->set('key', 'value', Ttl::forever());
```
Good point, @samdark!

Here we don’t call `->toSeconds()` for `Ttl::forever()` because it’s just a convenient shorthand for “infinite TTL” in PSR-16, which is represented by null. Similarly, `Ttl::from(null)` explicitly declares our intention of no expiration.

However, in a dynamic scenario where the TTL value might come from runtime code and could be `Ttl|DateInterval|int|null`, we do need to call `?->toSeconds()` to normalize it:

```php
$ttlFromCode = getTtlFromYourCode(); // Ttl|DateInterval|int|null
$cache->set('key', 'value', Ttl::from($ttlFromCode)?->toSeconds());
```


This ensures PSR-16 always receives a proper integer or null.

vjik commented on src/Ttl.php:
```diff
@@ -0,0 +1,150 @@
+     *
+     * @psalm-return null Indicates infinite TTL.
+     */
+    public static function forever(): ?self
```
Can we keep flag "forever" into VO as public readonly property?

This will allow to avoid `null` usage and create more semantic code. Insteadof `if ($ttl === null)` will be `if ($ttl->isForever)`

What do you think about this idea?

Pekhov14 replied:
Good idea 👍 
I’ve updated Ttl::forever() to return an object with a public readonly bool $isForever flag.
This makes the code more semantic (if ($ttl->isForever)) while keeping PSR-16 compatibility via toSeconds().

vjik commented on src/Ttl.php:
```diff
@@ -0,0 +1,181 @@
+    private const INT_PLACEHOLDER_FOR_FOREVER_VALUE = 0;
+
+    private function __construct(
+        public readonly int $value,
+        public readonly bool $isForever = false
```
```suggestion
        public readonly int|null $value,
```

Null means TTL is forever.

May be all negative values normalize to zero?

vjik commented on src/Ttl.php:
```diff
@@ -0,0 +1,181 @@
+            + $hours * self::SECONDS_IN_HOUR
+            + $days * self::SECONDS_IN_DAY;
+
+        if ($totalSeconds < 0) {
+            throw new \InvalidArgumentException('TTL must be non-negative.');
+        }
```
Check it in constructor only

vjik commented on src/Ttl.php:
```diff
@@ -0,0 +1,181 @@
+            ->add($interval)
+            ->getTimestamp();
+
+        if ($seconds < 0) {
+            throw new \InvalidArgumentException('DateInterval must result in non-negative TTL.');
+        }
```
Check it in constructor only

vjik commented on src/Ttl.php:
```diff
@@ -0,0 +1,181 @@
+     */
+    public static function forever(): self
+    {
+        return new self(self::INT_PLACEHOLDER_FOR_FOREVER_VALUE, true);
```
```suggestion
        return new self(null);
```

Pekhov14 commented:
Normalize expired TTL values to 0

Currently, expiry() returns -1 for expired TTLs. After introducing `Ttl` normalization, all negative TTL values are mapped to 0 (expired immediately).

I have updated the test to expect 0 instead of -1:

```php
$this->assertSame(-1, $items['key']->expiry());
// to
$this->assertSame(0, $items['key']->expiry());
```

Rationale:
null = forever
0 = expired
>0 = active

This keeps TTL values consistent and clear.

BC impact:
- The external cache behavior (get(), has(), delete()) remains unchanged.
- Only the internal representation of expired TTLs changed from -1 → 0.
- Consumers relying on the raw expiry() value of -1 should be aware of this update.

vjik commented:
> * Only the internal representation of expired TTLs changed from -1 → 0.
> * Consumers relying on the raw expiry() value of -1 should be aware of this update.

So may be remove normalization to keep BC?

Pekhov14 commented:
Removed normalization of negative TTL values (no automatic conversion to 0). 
Negative TTLs are now treated as expired without additional adjustment.

Moved the `toExpiration()` logic from `ArrayCache` into `Ttl`. 
This makes `Ttl` the single source of truth for expiration calculation.

This refactoring centralizes expiration logic in `Ttl` and simplifies cache implementations by relying on `Ttl` for consistent TTL handling.

vjik commented on CHANGELOG.md:
```diff
@@ -2,7 +2,7 @@
 ## 3.1.1 under development
 
-- no changes in this release.
+- Chg #150: Add `Ttl` value object for working with time-to-live duration (@Pekhov14)
```
```suggestion
- New #150: Add `Ttl` value object for working with time-to-live duration (@Pekhov14)
```

vjik commented on src/Ttl.php:
```diff
@@ -0,0 +1,233 @@
+     */
+    public function toExpiration(int $now): int
+    {
+        return match ($this->status()) {
```
Seems, this is `ArrayCache` internal logic. Expiration "0" or "-1" in common case is strange.

Pekhov14 commented on src/Ttl.php:
```diff
@@ -0,0 +1,233 @@
+     */
+    public static function fromInterval(DateInterval $interval): self
+    {
+        $seconds = (new DateTime('@0'))
+            ->add($interval)
+            ->getTimestamp();
```
`DateTime::add()` already respects the invert flag of `DateInterval`, so there’s no need for an explicit 
```php
$base = new DateTime('@0');
if ($interval->invert) {
    $seconds = $base->sub($interval)->getTimestamp();
} else {
    $seconds = $base->add($interval)->getTimestamp();
}
```
The shorter version with just `add($interval)` is correct and simpler.

vjik commented on README.md:
````diff
@@ -66,6 +66,102 @@ Set a default TTL:
+
+### Creating and Normalizing TTL
+
+The Ttl::from() method normalizes various TTL representations (Ttl, DateInterval, int, string, or null) into a Ttl object.
+```php
````
```suggestion
The `Ttl::from()` method normalizes various TTL representations (`Ttl`, `DateInterval`, `int`, `string`, or `null`) into a `Ttl` object.

```php
```

vjik commented on README.md:
````diff
@@ -66,6 +66,102 @@ Set a default TTL:
+
+- You can pass a `Ttl` object in the constructor as the default value.
+- You can pass it to methods like `getOrSet()` which expect integer number of seconds or `null`.
+```php
````
```suggestion

```php
```

vjik commented on README.md:
````diff
@@ -66,6 +66,102 @@ Set a default TTL:
+// Custom TTL
+$cache->getOrSet('key2', fn() => 'value2', Ttl::seconds(30)->toSeconds());
+$cache->getOrSet('key3', fn() => 'value3', Ttl::forever()->toSeconds()); // No expiration
+```
+### Additional Features
+Checking Infinite TTL: 
````
```suggestion
```

### Additional Features

Checking Infinite TTL: 
```

vjik commented on README.md:
```diff
@@ -66,6 +66,102 @@ Set a default TTL:
+### Additional Features
+Checking Infinite TTL: 
+
+Use isForever() to check if a TTL represents "forever" (i.e., no expiration). It returns true when the TTL value is null.
```
```suggestion
Use `isForever()` to check if a TTL represents "forever" (i.e., no expiration). It returns true when the TTL value is null.
```

vjik commented on README.md:
`````diff
@@ -66,6 +66,102 @@ Set a default TTL:
+}
+````
+
+### Accessing TTL Value
+Use toSeconds() to get the TTL in seconds (int) or null for "forever". The public $value property can be accessed directly (e.g., Ttl::seconds(30)->value), but toSeconds() is preferred for clarity.
`````
```suggestion
### Accessing TTL Value

Use `toSeconds()` to get the TTL in seconds (`int`) or `null` for "forever". The public `$value` property can be accessed directly (e.g., `Ttl::seconds(30)->value`), but `toSeconds()` is preferred for clarity.
```

vjik commented on README.md:
````diff
@@ -66,6 +66,102 @@ Set a default TTL:
+$seconds = $ttl->value; // Also 60
+```
+
+### Invalid TTL values
+```php
````
```suggestion
### Invalid TTL values

```php
```
